arch/risc-v/src/common/riscv_debug.c

```diff
@@ -0,0 +1,381 @@
+ ****************************************************************************/
+
+void riscv_debug_init(void)
+{
+  union mcontrol mc;
+
+  /* Attach the debug exception handler */
+
+  irq_attach(RISCV_IRQ_BPOINT, riscv_debug_handler, NULL);
+
+  /* Clear the global trigger info & mark them as free slots */
+
+  memset(g_trigger_map, 0, sizeof(g_trigger_map));
+
+  /* Detect the support of NAPOT by trigger 0 */
+
+  WRITE_CSR(CSR_TSELECT, 0);
+
+  mc.reg = READ_CSR(CSR_TDATA1);
+
+  mc.match = 1;
+
+  /* Write it to tdata1 and read back
+   * to check if the NAPOT is supported
+   */
+
+  WRITE_CSR(CSR_TDATA1, mc.reg);
+  mc.reg = READ_CSR(CSR_TDATA1);
+
+  if (mc.match == 1)
+    {
+      g_support_napot = true;
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_debugpoint_add
+ ****************************************************************************/
+
+int up_debugpoint_add(int type, FAR void *addr, size_t size,
+                      debug_callback_t callback, FAR void *arg)
+{
+  int slot;
+  union mcontrol mc;
+  int ret = OK;
+  uintptr_t addr_napot;
+
+  /* Find a free slot */
+
+  slot = debug_find_slot(0, 0, 0);
+
+  if (slot >= 0)
+    {
+      /* Select the trigger */
+
+      WRITE_CSR(CSR_TSELECT, slot);
+
+      /* Fetch the current setting from tdata1 */
+
+      mc.reg = READ_CSR(CSR_TDATA1);
+
+      /* Configure trigger */
+
+      mc.m = 1;
+      mc.u = 1;
+      mc.hit = 0;
+      mc.dmode = 0;
+      mc.action = 0;
+
+      mc.execute = 0;
+      mc.load = 0;
+      mc.store = 0;
+
+      if (type == DEBUGPOINT_BREAKPOINT)
+        {
+          mc.execute = 1;
+        }
+      else if (type == DEBUGPOINT_WATCHPOINT_RO)
+        {
+          mc.load = 1;
+        }
+      else if (type == DEBUGPOINT_WATCHPOINT_WO)
+        {
+          mc.store = 1;
+        }
+      else if (type == DEBUGPOINT_WATCHPOINT_RW)
+        {
+          mc.load = 1;
+          mc.store = 1;
+        }
+      else
+        {
+          ret = -EINVAL;
+          return ret;
+        }
+
+      /* From RISC-V Debug Specification:
+       * tdata1(mcontrol) match = 0 : Exact byte match
+       *
+       * tdata1(mcontrol) match = 1 : NAPOT (Naturally Aligned Power-Of-Two):
+       *
+       * Examples for understanding how to calculate match pattern to tdata2:
+       *
+       * nnnn...nnnnn 1-byte  Exact byte match
+       * nnnn...nnnn0 2-byte  NAPOT range
+       * nnnn...nnn01 4-byte  NAPOT range
+       * nnnn...nn011 8-byte  NAPOT range
+       * nnnn...n0111 16-byte NAPOT range
+       * nnnn...01111 32-byte NAPOT range
+       * ...
+       * n011...11111 2^31 byte NAPOT range
+       *  * where n are bits from original address
+       */
+
+      if (size > 1 && g_support_napot)
```

Comment:
why not use sizelo/sizehi

Reply:
sizelo/sizehi are relatively advanced feature, may not be implemented on many SoC